core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala

```diff
@@ -177,7 +177,7 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
     packagesExclusions = Option(packagesExclusions)
       .orElse(sparkProperties.get("spark.jars.excludes")).orNull
     deployMode = Option(deployMode).orElse(env.get("DEPLOY_MODE")).orNull
-    numExecutors = Option(numExecutors)
+    numExecutors = Option(numExecutors).orElse(env.get("SPARK_EXECUTOR_INSTANCES"))
```

Comment:
AFAIK, configuration property take high priority than System environment, here you just make the opposite.